src/ghci/mod.rs

```diff
@@ -0,0 +1,351 @@
+    /// The handle for the stdin interaction task.
+    stdin: JoinHandle<miette::Result<()>>,
+    /// A channel for sending events to interact with the stdin task.
+    stdin_channel: mpsc::Sender<StdinEvent>,
+    /// A channel for sending events to interact with the stdout task.
+    stdout_channel: mpsc::Sender<StdoutEvent>,
+    /// Count of 'sync' events sent. This lets us sync stdin/stdout -- we write a message to stdin
+    /// instructing `ghci` to print a sentinel string, and wait to read that string on `stdout`.
+    sync_count: AtomicUsize,
+    /// The currently-loaded modules in this `ghci` session.
+    modules: ModuleSet,
+}
+
+impl Debug for Ghci {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("Ghci").field(&self.process).finish()
+    }
+}
+
+impl Ghci {
+    /// Start a new `ghci` session using the given `command` to start `ghci`.
+    ///
+    /// This starts a number of asynchronous tasks to manage the `ghci` session's input and output
+    /// streams.
+    #[instrument(skip_all, level = "debug", name = "ghci")]
+    pub async fn new(command_arc: Arc<Mutex<Command>>) -> miette::Result<Arc<Mutex<Self>>> {
+        let mut child = {
+            let mut command = command_arc.lock().await;
+
+            command
+                .stdin(Stdio::piped())
+                .stderr(Stdio::piped())
+                .stdout(Stdio::piped())
+                .kill_on_drop(true);
+
+            command.spawn().into_diagnostic().wrap_err_with(|| {
+                format!("Failed to start `{}`", crate::command::format(&command))
+            })?
+        };
+
+        let stdin = child.stdin.take().unwrap();
+        let stdout = child.stdout.take().unwrap();
+        let stderr = BufReader::new(child.stderr.take().unwrap());
+
+        // TODO: Is this a good capacity? Maybe it should just be 1.
+        let (stdin_sender, stdin_receiver) = mpsc::channel(8);
+        let (stdout_sender, stdout_receiver) = mpsc::channel(8);
+
+        // So we want to put references to the `Ghci` struct we return in our tasks, but we don't
+        // have that struct yet. So we create some trivial tasks to construct a valid `Ghci`, and
+        // then create weak pointers to it and swap out the tasks.
+        let stdout_handle = task::spawn(async { Ok(()) });
+        let stderr_handle = task::spawn(async { Ok(()) });
+        let stdin_handle = task::spawn(async { Ok(()) });
+
+        let ret = Arc::new(Mutex::new(Ghci {
+            command: command_arc,
+            process: child,
+            stdout: stdout_handle,
+            stderr: stderr_handle,
+            stdin: stdin_handle,
+            stdin_channel: stdin_sender.clone(),
+            stdout_channel: stdout_sender.clone(),
+            sync_count: AtomicUsize::new(0),
+            modules: Default::default(),
+        }));
+
+        let (init_sender, init_receiver) = oneshot::channel::<()>();
```

Comment:
Yeah, I could essentially inline the `GhciStdout` state in here, but it would get pretty messy. It's mostly set up this way because the `GhciStdout` task owns the stdout reader.